csrc/ops/indexing.cpp

```diff
@@ -0,0 +1,268 @@
+      dim >= 0 && dim < (int)self_dom.size(),
+      "Scatter on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      self_dom.size(),
+      " non-reduction dims.");
+
+  // The shape of output tensor is same as self tensor.
+  std::vector<IterDomain*> out_domain;
+  for (const auto i : c10::irange(self_dom.size())) {
+    out_domain.push_back(
+        IterDomainBuilder(self_dom[i])
+            .iter_type(
+                self_dom[i]->getIterType() == IterType::Iteration
+                    ? IterType::GatherScatter
+                    : self_dom[i]->getIterType())
+            .build());
+  }
+
+  TensorView* out_tensor = IrBuilder::create<TensorView>(
+      IrBuilder::create<TensorDomain>(
+          out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
+      self->getDataType().value());
+
+  IrBuilder::create<ScatterOp>(
+      type, out_tensor, self, dim, index, src, out_domain[dim]);
+  return out_tensor->as<TensorView>();
+}
+
+TensorView* scatter(
+    TensorView* self,
+    int dim,
+    TensorView* index,
+    TensorView* src) {
+  return scatterOp(ScatterOpType::Set, self, dim, index, src);
+}
+
+TensorView* take_along_axis(TensorView* inp, TensorView* index, int64_t dim) {
+  const auto inp_domain =
+      TensorDomain::noReductions(inp->getMaybeRFactorDomain());
+  const auto idx_domain =
+      TensorDomain::noReductions(index->getMaybeRFactorDomain());
+
+  TORCH_CHECK(
+      !inp_domain.empty(), "take_along_axis can not be applied to 0d tensor.");
+  TORCH_CHECK(
+      idx_domain.size() == inp_domain.size(),
+      "The input and index tensor must have the same dimensions for take_along_axis");
+
+  if (dim < 0) {
+    dim += (int)idx_domain.size();
+  }
+
+  TORCH_CHECK(
+      dim >= 0 && dim < (int)inp_domain.size(),
+      "take_along_axis on invalid axis, received: ",
+      dim,
+      " however tensor view only has ",
+      inp_domain.size(),
+      " non-reduction dims.");
+
+  std::vector<IterDomain*> out_domain(idx_domain.size());
+
+  for (const auto i : c10::irange(idx_domain.size())) {
+    auto inp_id = inp_domain.at(i);
+    auto idx_id = idx_domain.at(i);
+
+    // Both inp_id and idx_id can be a broadcast domain, however,
+    // since the extent of inp_id cannot be smaller than the extent of
+    // idx_id, it doesn't make sense to have a broadcast input id and
+    // a non-broadcast index id. Specifically, when the extent of the
+    // input ID is one, the index extent must also be one.
+    if (inp_id->getMaybeExpandedExtent()->isOneInt()) {
+      TORCH_CHECK(
+          idx_id->getMaybeExpandedExtent()->isOneInt(),
+          "Invalid combination of input (",
+          inp_id->toString(),
+          ") and index domains (",
+          idx_id->toString(),
+          "). Input extent is one, but the index extent may be larger than one.");
+    }
+
+    out_domain.at(i) =
+        IterDomainBuilder(static_cast<int64_t>(i) == dim ? idx_id : inp_id)
```

Comment:
So for `gather`, we are setting almost every iterdomain as `IterType::GatherScatter`, but we are not doing so for `take_along_axis`? I don't think this is a problem, but I don't like this inconsistency

Reply:
It shouldn't be necessary, and I'm planning to remove its use from `torch_gather` as well. Will create an issue to track.

Reply:
https://github.com/NVIDIA/Fuser/issues/211